File: src/components/engine/logger.lua
---@meta logger

---
---@class Jila_Logger
---@field name string Jila_Logger name
Jila_Logger = {}

--- The Jila_Logger constructor
---@param name string Jila_Logger name
---@return Jila_Logger
function Create_Logger(name) end

---@param message string Message
function Jila_Logger:Warn(message) end

---@param message string Message
function Jila_Logger:Error(message) end

---@param message string Message
function Jila_Logger:Debug(message) end

---@param message string Message
function Jila_Logger:Info(message) end

File: src/engine/logger.cpp
#include <map>
#include "ulog.h"
#include "engine/logger.hpp"

namespace Jila {

Logger::Logger(std::string name) : name(name) {}

Logger::Logger() {}

void Logger::warn(std::string_view message) {
    ulog_warn("[%s] %s", name.c_str(), message.data());
}

void Logger::error(std::string_view message){
    ulog_error("[%s] %s", name.c_str(), message.data());
}

void Logger::debug(std::string_view message){
    ulog_debug("[%s] %s", name.c_str(), message.data());
}

void Logger::info(std::string_view message){
    ulog_info("[%s] %s", name.c_str(), message.data());
}

static std::map<std::string, Logger> loggers = {};

Logger& Logger::named(std::string_view name) {
    auto logger = loggers.find(name.data());

    if (logger == loggers.end()) {
        loggers[name.data()] = Logger(name.data());
    }

    return loggers[name.data()];
}

}
